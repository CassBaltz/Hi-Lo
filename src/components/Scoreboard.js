import React from "react";
import { connect } from "react-redux";
import { withStyles } from "@material-ui/styles";
import { PLAYER_ONE, PLAYER_TWO, SWITCH_PLAYER, } from "../redux/constants";

const DownArrow = ({ classes }) => (
  <svg
    className={classes.arrow}
    xmlns="http://www.w3.org/2000/svg"
    height="24"
    viewBox="0 0 24 24"
    width="24"
  >
    <path d="M7 10l5 5 5-5z" />
    <path d="M0 0h24v24H0z" fill="none" />
  </svg>
);

const UpArrow = ({ classes }) => (
  <svg
    className={classes.arrow}
    xmlns="http://www.w3.org/2000/svg"
    height="24"
    viewBox="0 0 24 24"
    width="24"
  >
    <path d="M7 14l5-5 5 5z" />
    <path d="M0 0h24v24H0z" fill="none" />
  </svg>
);

const GAME_STATUSES = {
  TIED: "TIED",
  WINNING: "WINNING",
  LOSING: "LOSING",
};

const Scoreboard = ({ classes, playerOne, playerTwo, }) => {
  return (
    <div className={classes.root}>
      <div className={`${classes.playerWrapper} ${classes.playerOne}`}>
        <p>{playerOne.name}</p>
        <p>{playerOne.score}</p>
      </div>
      <div className={classes.divider}>{playerOne.isPlaying ? <UpArrow classes={classes} /> : <DownArrow classes={classes} />}</div>
      <div className={`${classes.playerWrapper} ${classes.playerTwo}`}>
        <p>{playerTwo.name}</p>
        <p>{playerTwo.score}</p>
      </div>
    </div>
  )
};


const styles = {
  root: {
    display: "flex",
    flexDirection: "column",
    border: "2px solid #adb5bd",
    borderRadius: "3px",
    backgroundColor: "#FFF",
  },
  playerWrapper: {
    display: "flex",
    alignItems: "center",
    justifyContent: "space-between",
    "& p": {
      margin: 0,
    },
    "& p:last-child": {
      marginLeft: "20px",
    },
    padding: "10px",
  },
  divider: {
    display: "flex",
    justifyContent: "center",
  },
  arrow: {
    fill: "#adb5bd",
  },
  playerOne: {
    color: props => getStatusColor(props.playerOne),
  },
  playerTwo: {
    color: props => getStatusColor(props.playerTwo),
  }
};

const getStatusColor = player => {
  switch (player.gameStatus) {
    case GAME_STATUSES.TIED:
      return "#ffd43b";
    case GAME_STATUSES.LOSING:
      return "#e03131";
    case GAME_STATUSES.WINNING:
    default:
      return "#37b24d";
  }
};

const mapStateToProps = ({ game }) => {
  const playerOne = { ...game[PLAYER_ONE] };
  const playerTwo = { ...game[PLAYER_TWO] };
  playerOne.gameStatus = getGameStatus(game, PLAYER_ONE);
  playerOne.isPlaying = game.turn.player === PLAYER_ONE;
  playerTwo.gameStatus = getGameStatus(game, PLAYER_TWO);

  return {
    playerOne,
    playerTwo,
  };
};

export const getGameStatus = (game, player) => {
  const thisPlayerScore = game[player].score;
  const opponentScore = game[SWITCH_PLAYER[player]].score;

  if (thisPlayerScore < opponentScore) {
    return GAME_STATUSES.WINNING;
  }

  if (thisPlayerScore > opponentScore) {
    return GAME_STATUSES.LOSING;
  }

  return GAME_STATUSES.TIED;
}

export default connect(mapStateToProps)(withStyles(styles)(Scoreboard));