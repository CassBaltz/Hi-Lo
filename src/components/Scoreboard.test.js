import React from "react";

import { Scoreboard, styles, } from "./Scoreboard";

const props = {
  classes: classReducer(styles),
  playerOne: {
    name: "mike",
    score: 0,
    isPlaying: true,
  },
  playerTwo: {
    name: "jim",
    score: 0,
  },
};


describe("<Scoreboard />", () => {
  it("should render correctly when given the default props", () => {
    const scoreboard = shallow(<Scoreboard {...props} />);

    expect(toJson(scoreboard)).toMatchSnapshot();
  });

  it("should render correctly when player one is winning", () => {
    const playerOneWinningProps = {
      ...props,
      playerTwo: {
        ...props.playerTwo,
        score: 1,
      }
    };
    const scoreboard = shallow(<Scoreboard {...playerOneWinningProps} />);

    expect(toJson(scoreboard)).toMatchSnapshot();
  });

  it("should render correctly when player two is winning and it is player two's turn", () => {
    const playerTwoWinningProps = {
      ...props,
      playerOne: {
        ...props.playerOne,
        score: 1,
        isPlaying: false,
      }
    };
    const scoreboard = shallow(<Scoreboard {...playerTwoWinningProps} />);

    expect(toJson(scoreboard)).toMatchSnapshot();
  });
});