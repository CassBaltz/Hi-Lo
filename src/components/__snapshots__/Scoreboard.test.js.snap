// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Scoreboard /> should render correctly when given the default props 1`] = `
<div
  className="root"
>
  <div
    className="playerWrapper playerOne"
  >
    <p>
      mike
    </p>
    <p>
      0
    </p>
  </div>
  <div
    className="divider"
  >
    <UpArrow
      classes={
        Object {
          "arrow": "arrow",
          "divider": "divider",
          "playerOne": "playerOne",
          "playerTwo": "playerTwo",
          "playerWrapper": "playerWrapper",
          "root": "root",
        }
      }
    />
  </div>
  <div
    className="playerWrapper playerTwo"
  >
    <p>
      jim
    </p>
    <p>
      0
    </p>
  </div>
</div>
`;

exports[`<Scoreboard /> should render correctly when player one is winning 1`] = `
<div
  className="root"
>
  <div
    className="playerWrapper playerOne"
  >
    <p>
      mike
    </p>
    <p>
      0
    </p>
  </div>
  <div
    className="divider"
  >
    <UpArrow
      classes={
        Object {
          "arrow": "arrow",
          "divider": "divider",
          "playerOne": "playerOne",
          "playerTwo": "playerTwo",
          "playerWrapper": "playerWrapper",
          "root": "root",
        }
      }
    />
  </div>
  <div
    className="playerWrapper playerTwo"
  >
    <p>
      jim
    </p>
    <p>
      1
    </p>
  </div>
</div>
`;

exports[`<Scoreboard /> should render correctly when player two is winning and it is player two's turn 1`] = `
<div
  className="root"
>
  <div
    className="playerWrapper playerOne"
  >
    <p>
      mike
    </p>
    <p>
      1
    </p>
  </div>
  <div
    className="divider"
  >
    <DownArrow
      classes={
        Object {
          "arrow": "arrow",
          "divider": "divider",
          "playerOne": "playerOne",
          "playerTwo": "playerTwo",
          "playerWrapper": "playerWrapper",
          "root": "root",
        }
      }
    />
  </div>
  <div
    className="playerWrapper playerTwo"
  >
    <p>
      jim
    </p>
    <p>
      0
    </p>
  </div>
</div>
`;
